internal/controller/instance/delta/delta.go

```diff
@@ -0,0 +1,181 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package delta
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// Difference represents a single field-level difference between two objects.
+// Path is the full path to the differing field (e.g. "spec.containers[0].image")
+// Desired and Observed contain the actual values that differ at that path.
+type Difference struct {
+	// Path is the full path to the differing field (e.g. "spec.x.y.z"
+	Path string `json:"path"`
+	// Desired is the desired value at the path
+	Desired interface{} `json:"desired"`
+	// Observed is the observed value at the path
+	Observed interface{} `json:"observed"`
+}
+
+// Compare takes desired and observed unstructured objects and returns a list of
+// their differences. It performs a deep comparison while being aware of Kubernetes
+// metadata specifics. The comparison:
+//
+// - Cleans metadata from both objects to avoid spurious differences
+// - Walks object trees in parallel to find actual value differences
+// - Builds path strings to precisely identify where differences occurs
+// - Handles type mismatches, nil values, and empty vs nil collections
+func Compare(desired, observed *unstructured.Unstructured) ([]Difference, error) {
+	desiredCopy := desired.DeepCopy()
+	observedCopy := observed.DeepCopy()
+
+	cleanMetadata(desiredCopy)
+	cleanMetadata(observedCopy)
+
+	var differences []Difference
+	walkCompare(desiredCopy.Object, observedCopy.Object, "", &differences)
+	return differences, nil
+}
+
+// ignoredFields are Kubernetes metadata fields that should not trigger updates.
+var ignoredFields = []string{
```

Comment:
Will we add more fields to ignore here? or just the ones in metadata?
Because instead i'm thinking we can call it ignoredMetadataFields...